weather_aplication/MainViewController.swift

```diff
@@ -1,9 +1,123 @@
+        dataArray.append(CellData(hour: "20:00", temperature: 31, pressure: 1011, windSpeed: 15))   //11
+        dataArray.append(CellData(hour: "21:00", temperature: 32, pressure: 1012, windSpeed: 16))   //12
+    }
     
     override func viewDidLoad() {
+        super.viewDidLoad()
+        createDataArray()
+        setupView()
+        setupConstraints()
+        
+    }
+    
+    func setupView() {
         view.backgroundColor = .lightGray
+        
+        cityLabel.textAlignment = .center
+        cityLabel.text = cityString
+        view.addSubview(cityLabel)
+        
+        weatherTableView.dataSource = self
+        weatherTableView.delegate  = self
+        weatherTableView.backgroundColor = .white
+        view.addSubview(weatherTableView)
+        weatherTableView.register(CustomTableViewCell.self, forCellReuseIdentifier: "cell")
+    }
+    
+    func setupConstraints() {
+        
+        cityLabel.snp.makeConstraints { make in
+            make.top.equalTo(view.snp.topMargin)
+            make.leftMargin.equalToSuperview().offset(kSideMargin)
+            make.rightMargin.equalToSuperview().offset(-kSideMargin)
+            make.centerX.equalTo(view.snp.centerX)
+        }
+        
+        weatherTableView.snp.makeConstraints { make in
+            make.top.equalTo(cityLabel.snp.bottom).offset(kTopMargin)
+            make.leftMargin.rightMargin.bottomMargin.equalToSuperview()
+        }
+    }
+}
+
+extension HourlyWeatherViewController: UITableViewDelegate, UITableViewDataSource {
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+      return dataArray.count
+    }
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        guard let cell = tableView.dequeueReusableCell(withIdentifier: "cell", for: indexPath as IndexPath) as? CustomTableViewCell else {fatalError("Unble to create cell")}
+        cell.hourLabel.text = dataArray[indexPath.row].hour
+        cell.temperatureLabel.text = "\(dataArray[indexPath.row].temperature)C"
+        cell.pressureLabel.text = "\(dataArray[indexPath.row].pressure)hPa"
+        cell.windSpeedLabel.text = "\(dataArray[indexPath.row].windSpeed)km/h"
+        return cell
+    }
+}
+
+struct CellData {
+    lazy var hour = String()
+    lazy var temperature = Int()
+    lazy var pressure = Int()
+    lazy var windSpeed = Int()
+}
+
+class CustomTableViewCell: UITableViewCell {
+    private let kSideMargin: CGFloat = 20
+    lazy var hourLabel = UILabel()
+    lazy var temperatureLabel = UILabel()
+    lazy var pressureLabel = UILabel()
+    lazy var windSpeedLabel = UILabel()
+    
+    override func setSelected(_ selected: Bool, animated: Bool) {
+        super.setSelected(selected, animated: animated)
+        addSubview(hourLabel)
+        addSubview(temperatureLabel)
+        addSubview(pressureLabel)
+        addSubview(windSpeedLabel)
+    }
+    override func layoutSubviews() {
```

Comment:
Tej funkcji nie powinniśmy nadpisywać w takim celu jak to tutaj jest zrobione, przeczytaj sobie w dokumentacji i napisz mi dlaczego nie powinniśmy tego robić. Zastanów się też w jaki sposób zamiast tego powinniśmy dodawać subviewsy i constrainty

Reply:
zamiast tego powinna zostać użyta func setNeedsLayout()

Reply:
No nie do końca, layoutIfNeeded jak sama nazwa wskazuje mówi widokowi że powinien odświeżyć to co pokazuje na ekranie. Tyle że ta funkcja jest wywoływana za każdym razem gdy widok będzie się pokazywać na ekranie, podobnie layoutSubviews. Na dodatek sama cellki są reużywane- możesz sobie o tym doczytać jak ten mechanizm działa. Właśnie z tego powodu konkretne widoki będą nawet kilka razy dodawane jako subview. Dlatego też musimy funkcje setupView i setupConstraints dodać w miejscu które jest tylko raz wywoływane. I takim miejscem jest właśnie init cellki, który będziesz musiał nadpisać